core/jobs/batch_jobs/voiceover_migration_job.py

```diff
@@ -0,0 +1,358 @@
+                content_id_to_voiceovers_mapping.items()):
+            for lang_code, voiceover_mapping in (
+                    lang_code_to_voiceover_mapping.items()):
+                voice_artist_id = voiceover_mapping[0]
+                voiceover_dict = voiceover_mapping[1]
+
+                try:
+                    accent_code = voice_artist_id_to_language_code_mapping[
+                        voice_artist_id][lang_code]
+                except Exception as e:
+                    raise Exception(
+                        'Please assign all the accents for voice artists in '
+                        'language code %s.' % lang_code) from e
+
+                entity_voiceover_id_to_entity_voiceovers = (
+                    cls.update_entity_voiceover_for_given_id(
+                        entity_type,
+                        entity_id,
+                        entity_version,
+                        accent_code,
+                        content_id,
+                        voiceover_dict,
+                        entity_voiceover_id_to_entity_voiceovers
+                    )
+                )
+        return list(entity_voiceover_id_to_entity_voiceovers.values())
+
+    @classmethod
+    def count_voiceovers(
+        cls,
+        entity_voiceovers_model: voiceover_models.EntityVoiceoversModel
+    ) -> int:
+        """Calculates the number of voiceovers for a given entity
+        voiceover model.
+
+        Args:
+            entity_voiceovers_model: EntityVoiceoversModel. An instance of an
+                entity voiceover model from which the number of voiceovers will
+                be calculated.
+
+        Returns:
+            int. The number of voiceovers present in the given entity
+            voiceovers model.
+        """
+        total_voiceovers: int = 0
+        for voiceover_type_to_voiceovers in (
+                entity_voiceovers_model.voiceovers.values()):
+            if 'manual' in voiceover_type_to_voiceovers:
+                total_voiceovers += 1
+        return total_voiceovers
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of results for the exploration for which an
+        exploration voiceover migration has been done.
+
+        Returns:
+            PCollection. A PCollection of results for the exploration for which
+            an exploration voiceover migration has been done.
+        """
+        exploration_models = (
+            self.pipeline
+            | 'Get exploration models' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all())
+        )
+
+        # Pair each exploration model with its exploration ID and create
+        # key-value pairs.
+        paired_exploration_models = (
+            exploration_models
+            | 'Pair Exploration ID to model' >> beam.Map(
+                lambda model: (model.id, model))
+        )
+
+        exploration_voice_artists_link_models = (
```

Comment:
Isn't the usual paradigm to do all the gets at the start, then do all the jobs, then do all the puts?